### PR TITLE
#34913 Support for sparse configs

### PR DESCRIPTION
**Pull request by josh-t**
These changes eliminate the requirement that all bundle settings must be explicitly specified in an environment. The new behavior uses the `default_value` specified in the bundle's `info.yml` manifest as a fallback at runtime when not specified in the environment file.

Any `tank` commands (`update`, `install`, etc) that modify environment files will no longer populate new settings and set their values to the manifest's corresponding default value. Settings with no default value will still be added to the environment after prompting the user for a value. It should be pointed out that no existing settings will be removed from environment files with the changes.

### CODE REVIEW

File: python/tank/platform/bundle.py [hunk 1]
```diff
@@ -823,9 +821,28 @@ def __resolve_setting_value(self, key, value):
...
             schema = None
-        
+
         if schema:
-            # post process against schema
-            value = self.__post_process_settings_r(key, value, schema)
+
+            # If the value is None, then the setting wasn't specified. Use the
+            # default value in the schema instead.
+            if value is None:
```
Comment (robblau) [inline python/tank/platform/bundle.py]: What happens if None is a valid value?  Is there a way to know if the setting is actually missing?

File: python/tank/platform/bundle.py [hunk 2]
```diff
@@ -823,9 +821,28 @@ def __resolve_setting_value(self, key, value):
...
+
+                if schema.get("type") == "hook":
+                    # Hooks already have the concept of a default value. Just
+                    # use that special default hook string and allow the post
+                    # process to resolve the value correctly based on the
+                    # default value in the manifest. Because the hooks have a
+                    # bit of legacy path evaluation code, if we just use the
+                    # value from the manifest here, we aren't guaranteed to get
+                    # the expected path.
+                    value = constants.TANK_BUNDLE_DEFAULT_HOOK_SETTING
```
Comment (robblau) [inline python/tank/platform/bundle.py]: If the default value in the info.yml isn't this, wouldn't this do the wrong thing?

File: python/tank/platform/validation.py [hunk 3]
```diff
@@ -285,9 +285,15 @@ def __validate_schema_value(self, settings_key, schema):
         data_type = schema.get("type")
         self.__validate_schema_type(settings_key, data_type)
 
-        if "default_value" in schema:
+        # Get a list of all keys that start with the default value key string.
+        # Some types, like hooks, allow for engine specific default values such
+        # as "default_value_tk-maya". Validate each of these key's values
+        # against the setting's data type.
+        default_value_keys = [k for k in schema if k.startswith("default_value")]
```
Comment (robblau) [inline python/tank/platform/validation.py]: Shouldn't this test be setting type specific?  Since this would only apply to hooks?  There would be an edge case if you had a string that started with "default_value", wouldn't you?

File: python/tank/platform/validation.py [hunk 4]
```diff
@@ -379,16 +389,31 @@ def validate(self, settings):
...
+            if settings_key in settings:
+                # value exists in the settings. use it.
+                settings_value = settings[settings_key]
+            else:
+                # No value in the settings. Look for a default value.
+                if value_schema.get("type") == "hook":
+                    # Hooks already have the concept of a default value. If this
+                    # setting is a hook, just use that special default hook
+                    # string and allow the validation code do its thing.
+                    settings_value = constants.TANK_BUNDLE_DEFAULT_HOOK_SETTING
```
Comment (robblau) [inline python/tank/platform/validation.py]: Same as above, shouldn't the value from the manifest be used since it could be different than the constant?

File: python/tank/platform/validation.py [hunk 5]
```diff
@@ -379,16 +389,31 @@ def validate(self, settings):
...
+                else:
+                    settings_value = value_schema.get("default_value", None)
+
+            # if no value, then can't validate the setting. raise.
+            if settings_value is None:
```
Comment (robblau) [inline python/tank/platform/validation.py]: Is None an invalid value?
Shouldn't the test be if there isn't a value in the env and there isn't a default for the setting in the manifest?

File: python/tank/platform/bundle.py (hunk 1 above)
Comment (josh-t) [inline python/tank/platform/bundle.py]: Yeah, I think you're right. I think it should look to see if `key` is in `self.__settings` instead. If it is, then proceed with whatever the value is (`None` or otherwise). If not, then it was omitted and it should try to determine the default.

File: python/tank/platform/bundle.py (hunk 2 above)
Comment (josh-t) [inline python/tank/platform/bundle.py]: From what I could tell, this setting triggers essentially the sparse config behavior (lookup the default value at runtime). If we just go with what's in the `info.yml` file, then the path evaluation doesn't work as (currently) expected. I think we need this for backward compatibility. I'll check with @manneohrstrom to verify my thought process.

File: python/tank/platform/validation.py (hunk 3 above)
Comment (josh-t) [inline python/tank/platform/validation.py]: From what I could tell in the code, these special `default_value_{engine_name}` keys aren't specific to hooks types. Please correct me if I'm wrong. 

FYI, This particular section didn't technically need to be modified for sparse configs, I just noticed that it wasn't validating the engine-specific default values. So I guess this could cause breakage for folks with invalid, but never used, engine-specific defaults in their manifests. Let me know if you want to not do this.

File: python/tank/platform/validation.py (hunk 4 above)
Comment (josh-t) [inline python/tank/platform/validation.py]: If we use the value from the manifest, we'll need to modify all the manifests to use the "{self}" form instead of just the hook name. I tried to fall back directly to the manifest originally, and the apps failed because they were looking for hooks in the config's `hooks` directory since the legacy defaulting of the path wasn't kicking in. It would be good to get @manneohrstrom to have a look here as well.

File: python/tank/platform/validation.py (hunk 5 above)
Comment (josh-t) [inline python/tank/platform/validation.py]: Yep. I will refactor this.

File: python/tank/platform/validation.py (hunk 3 above)
Comment (manneohrstrom) [inline python/tank/platform/validation.py]: Yepp, it can be any type! I think we introduced that syntax for the loader, that needed different defaults for different engines. Example here: https://github.com/shotgunsoftware/tk-multi-loader2/blob/master/info.yml#L60

File: python/tank/platform/bundle.py [hunk 6]
```diff
@@ -804,28 +804,57 @@ def __post_process_settings_r(self, key, value, schema):
...
-        # (may fail if the key does not exist in the schema,
-        # which is an old use case we need to support now...)
+
+        # start with the supplied default value in case we need to short circuit
+        # before checking against the schema
+        value = default
+
+        # attempt to get the value from the settings.
+        if key in settings:
+            value = settings[key]
```
Comment (manneohrstrom) [inline python/tank/platform/bundle.py]: aren't we done at this point? Would it make sense to do

```python
if key in settings:
    # a setting was explicitly specified in the environment config 
    value = settings[key]
else:
    # no setting specified in the environment config, look up the default
    # from the bundle manifest
    rest of code here....

File: python/tank/platform/bundle.py [hunk 7]
```diff
@@ -804,28 +804,57 @@ def __post_process_settings_r(self, key, value, schema):
...
-        if schema:
-            # post process against schema
-            value = self.__post_process_settings_r(key, value, schema)
-            
-        return value
+
+        if not schema:
+            # No schema, can't post process or retrieve a default. Just return
+            # the value we have.
+            return value
```
Comment (manneohrstrom) [inline python/tank/platform/bundle.py]: would this be clearer if it just did `return default`?

File: python/tank/platform/bundle.py [hunk 8]
```diff
@@ -804,28 +804,57 @@ def __post_process_settings_r(self, key, value, schema):
...
+
+        # Schema exists. If the key wasn't in the settings, fallback to the
+        # default value in the manifest.
+        if key not in settings:
```
Comment (manneohrstrom) [inline python/tank/platform/bundle.py]: if this is instead part of a bigger if...else clause - something like i suggested above, then maybe we can skip this check? Could it be something like 

```
if key in settings:
    # a setting was explicitly specified in the environment config 
    value = settings[key]
else:
    # no setting specified in the environment config, look up the default
    # from the bundle manifest
    schema = try_to_get_schema()
    if not schema:
        # no schema info for requested param
        value = default
    else:
        # pull default from schema 
        value = schema.get("default_value", default)

# post process non-empty
if value:
    value = self.__post_process_settings_r(key, value, schema)

return value
```

File: python/tank/platform/bundle.py [hunk 9]
```diff
@@ -804,28 +804,57 @@ def __post_process_settings_r(self, key, value, schema):
...
+                # use that special default hook string and allow the post
+                # process to resolve the value correctly based on the
+                # default value in the manifest. Because the hooks have a
+                # bit of legacy path evaluation code, if we just use the
+                # value from the manifest here, we aren't guaranteed to get
+                # the expected path.
+                value = constants.TANK_BUNDLE_DEFAULT_HOOK_SETTING
+            else:
+                # Use the default value from the manifest
+                value = schema.get("default_value", default)
```
Comment (manneohrstrom) [inline python/tank/platform/bundle.py]: this does not include the logic to cater for `default_value_tk-maya` - see https://github.com/shotgunsoftware/tk-core/blob/master/python/tank/deploy/tank_commands/console_utils.py#L505

Also, what happens with that logic - in https://github.com/shotgunsoftware/tk-core/blob/master/python/tank/deploy/tank_commands/console_utils.py#L505 - now that all those lookups happen at runtime? Can we remove it?

File: python/tank/platform/bundle.py [hunk 10]
```diff
@@ -804,28 +804,57 @@ def __post_process_settings_r(self, key, value, schema):
...
+        if not schema:
+            # No schema, can't post process or retrieve a default. Just return
+            # the value we have.
+            return value
+
+        # Schema exists. If the key wasn't in the settings, fallback to the
+        # default value in the manifest.
+        if key not in settings:
+
+            if schema.get("type") == "hook":
```
Comment (manneohrstrom) [inline python/tank/platform/bundle.py]: Could this value-resolve logic be broken out into a method and reused? It seems to be in two places and with the more complex logic for defaults (default_value_tk-maya), it will grow in size and complexity.

File: python/tank/platform/validation.py (hunk 3 above)
Comment (manneohrstrom) [inline python/tank/platform/validation.py]: The code here - https://github.com/shotgunsoftware/tk-core/blob/master/python/tank/deploy/tank_commands/console_utils.py#L505 - takes the engine name and uses that to compose a key default_value_xxx and looks for that - would it make sense to apply the same logic here?

File: python/tank/platform/validation.py [hunk 11]
```diff
@@ -379,16 +389,31 @@ def validate(self, settings):
...
+                    # string and allow the validation code do its thing.
+                    settings_value = constants.TANK_BUNDLE_DEFAULT_HOOK_SETTING
+                else:
+                    if "default_value" not in value_schema:
+                        # if no value, then can't validate the setting. raise.
+                        raise TankError(
+                            "Could not determine value for key '%s' in settings!"
+                            "No specified value and no default value." % settings_key)
+                    else:
+                        settings_value = value_schema.get("default_value", None)
```
Comment (manneohrstrom) [inline python/tank/platform/validation.py]: I was expecting to see more of the logic removed from within console utils to reappear here - what if the hook default value is `{self}/path/to/{engine_name}_foo.py` ?

**Comment by manneohrstrom** [comment]
Added some comments. Do we have unit tests for the default_value_tk-maya syntax? And the {engine_name} token for hooks. If not (which i am guessing is the answer) maybe this could be a good time to add that?

File: python/tank/platform/validation.py (hunk 11 above)
Comment (robblau) [inline python/tank/platform/validation.py]: Yes please.  It'll be so nice to have more of the settings code all in one place!

File: python/tank/platform/validation.py (hunk 4 above)
Comment (manneohrstrom) [inline python/tank/platform/validation.py]: I think we need to add that big chunk of logic that is currently in console_utils (where it writes a default value into the env) and apply that same logic here. 

The above code sets the value to "default" - This is the "old" hooks format. We then added "new style" hook syntax on top of that, where the value is something like `{self}/foo.py` or `{config}/foo.py` e.g. a format which is much more similar to a normal setting in the sense that it currently does get baked into the env at install time.

Comment (manneohrstrom) [inline python/tank/platform/validation.py]: Examples of this:

### old style
- manifest definition: https://github.com/shotgunsoftware/tk-multi-publish/blob/master/info.yml#L65
- installed in config: https://github.com/shotgunsoftware/tk-config-default/blob/master/env/shot_step.yml#L71
So here, tk will discover *default* look at the manifest value (`primary_publish`) and resolve the hook path `PATH_TO_BUNDLE/hooks/primary_publish.py` at runtime.

### new style
- manifest def: https://github.com/shotgunsoftware/tk-multi-loader2/blob/master/info.yml#L31
- in config: https://github.com/shotgunsoftware/tk-config-default/blob/master/env/shot_step.yml#L479

Here, we don't need to do the funny roundtrip into the manifest that we did with the old style hooks (e.g. to figure out that `default` actually means `primary_publish` in the case of the setting above). Instead, it's just baked out and much more readable in the config: `actions_hook: '{self}/tk-maya_actions.py'` - it is easy to see what tk will be doing (good transparency) and it's easy to configure, typically a good starting point it so just replace `{self}` with `{config}`.

Also note how in this case, the manifest def `default_value: "{self}/{engine_name}_actions.py"` has got `{engine_name}` resolved at install time but the rest of the tokens resolved at runtime.

File: python/tank/platform/validation.py (hunk 11 above)
Comment (manneohrstrom) [inline python/tank/platform/validation.py]: :+1: that would be awesome!

File: python/tank/platform/bundle.py (hunk 7 above)
Comment (josh-t) [inline python/tank/platform/bundle.py]: Well, it could be the value from the settings I think. I'm looking at your other comments to see about refactoring.

File: python/tank/platform/bundle.py (hunk 8 above)
Comment (josh-t) [inline python/tank/platform/bundle.py]: In your example, the value could be specified in the settings, but no schema is available. That would still cause the post processing to run which expects the schema. I'll go over this code again.

File: python/tank/platform/bundle.py (hunk 6 above)
Comment (josh-t) [inline python/tank/platform/bundle.py]: Looking at the code again to see about cleaning it up a bit.

File: python/tank/platform/bundle.py (hunk 10 above)
Comment (josh-t) [inline python/tank/platform/bundle.py]: Makes sense. Will do.

File: python/tank/platform/bundle.py (hunk 9 above)
Comment (josh-t) [inline python/tank/platform/bundle.py]: Ok. Yeah, I think I need to add/call the missing logic from the post processing method anyway. Will do.

File: python/tank/platform/validation.py (hunk 3 above)
Comment (josh-t) [inline python/tank/platform/validation.py]: The logic in that method is looking for a specific key given an engine name. This code is validating all the `default_value_{engine_name}` values in the schema (no engine name available). So I don't think there's logic to be shared here.

Comment (manneohrstrom) [inline python/tank/platform/validation.py]: gotcha! Sorry!

File: python/tank/platform/validation.py (hunk 4 above)
Comment (josh-t) [inline python/tank/platform/validation.py]: @manneohrstrom - This all makes sense. I'm still trying to wrangle backward compatibility. We need to still support `default` as a valid hook value in the environments, and we need to support `primary_publish` style hook defaults in the manifests...

With sparse configs, when the setting is asked for, it will fall back to the value in the manifest if not specified in the environment. For hooks, this means it would return say, `primary_publish` for an old-style hook whereas in the past it would have returned the magic `default` value added to the settings at install/update time. Now, when the hook is executed, the magic `default` path is resolved as the bundle's `hooks` directory, but `primary_publish` value from the manifest is resolved to the config's `hooks` directory. 

So there has to be something special that happens to old-style hook default values in order to maintain backward compatibility I think. So either return the magic `default` value or prepend `{self}`. Does that make sense? Maybe I'm overthinking things. 

I'll push my latest WIP shortly and maybe that'll help frame my thought process.

File: python/tank/deploy/tank_commands/console_utils.py [hunk 12]
```diff
@@ -190,95 +190,16 @@ def _get_configuration_recursive(log, tank_api_instance, new_ver_descriptor, par
...
-                            hooks_folder = os.path.join(new_ver_descriptor.get_path(), "hooks")
-                            full_path = resolved_default_value.replace("{self}", hooks_folder)
-                            
-                            if not os.path.exists(full_path):
-                                log.warning("Sorry, no built-in support for the %s engine yet! " 
-                                            "The default hook value '%s' for this hook refers to an engine specific hook, "
-                                            "however there is currently "
-                                            "no hook implementation available for the current engine. "
-                                            "If you want to continue with the install, you have to supply your "
-                                            "own hooks." % (parent_engine_name, default_value))
```
Comment (manneohrstrom) [inline python/tank/deploy/tank_commands/console_utils.py]: Hmm, so I think maybe the rationale here is that this check happens at install time, where you don't refer to any hooks other than `{self}`. The default value in an app manifest could technically be `{config}/foo/bar.py` but it wouldn't make any sense I think because it assumes certain things about the config you are about to install into! The only thing we know for sure exists is `{self}`. I think this is why the code only covers that case. Would have been nice with some comments in the existing code explaining this!

If we move this logic so that it is resolved at runtime, we should validate all syntaxes!

Comment (manneohrstrom) [inline python/tank/deploy/tank_commands/console_utils.py]: Hmm, so I think maybe the rationale here is that this check happens at install time, where you don't refer to any hooks other than {self}. The default value in an app manifest could technically be {config}/foo/bar.py but it wouldn't make any sense I think because it assumes certain things about the config you are about to install into! The only thing we know for sure exists is {self}. I think this is why the code only covers that case. Would have been nice with some comments in the existing code explaining this!

If we move this logic so that it is resolved at runtime, we should validate all syntaxes!

File: python/tank/deploy/tank_commands/console_utils.py [hunk 13]
```diff
@@ -482,38 +405,14 @@ def _generate_settings_diff_recursive(parent_engine_name, old_schema, new_schema
...
-            # special case handling for list params - check if
-            # allows_empty == True, in that case set default value to []
-            if (param_type == "list"
-                and new_params[param_name].get("value") == None
-                and new_param_definition_dict.get("allows_empty") == True):
-                new_params[param_name]["value"] = []
+            # attempt to resolve a default value from the new parameter def.
+            # Use the fallback default with an unlikely to be used value to
+            # detect cases where there is no default value in the schema.
+            no_default_value = "__NO_DEFAULT_VALUE_IN_SCHEMA__"
```
Comment (manneohrstrom) [inline python/tank/deploy/tank_commands/console_utils.py]: Instead of the `no_default_value`, could we pass in a `strict` flag or something to make the method raise an exception in the case a value is not found or something like that?

File: python/tank/deploy/tank_commands/validate_config.py [hunk 14]
```diff
@@ -125,7 +125,7 @@ def _run(self, log):
 g_templates = set()
 g_hooks = set()
 
-def _validate_bundle(log, tk, name, settings, descriptor):
+def _validate_bundle(log, tk, name, settings, descriptor, engine_name=None, app_name=None):
```
Comment (manneohrstrom) [inline python/tank/deploy/tank_commands/validate_config.py]: since you are touching this method, could you add doc strings? :)

File: python/tank/deploy/tank_commands/validate_config.py [hunk 15]
```diff
@@ -145,54 +145,54 @@ def _validate_bundle(log, tk, name, settings, descriptor):
...
-                if manifest[s].get("type") == "hook":
-                    g_hooks.add(value)
 
-                     
-    for r in manifest.keys():
-        if r not in settings.keys():
-            log.error("Required parameter missing: %s" % r)
+    for s in manifest.keys():
+
+        default = bundle.resolve_default_value(manifest.get(s), engine_name=engine_name)
```
Comment (manneohrstrom) [inline python/tank/deploy/tank_commands/validate_config.py]: maybe `manifest[s]` instead of `get()` to highlight that we expect a key to always exist?

Comment (manneohrstrom) [inline python/tank/deploy/tank_commands/validate_config.py]: is engine_name always tk-shell for the tank command? Is that potentially an issue?

File: python/tank/deploy/tank_commands/validate_config.py [hunk 16]
```diff
@@ -145,54 +145,54 @@ def _validate_bundle(log, tk, name, settings, descriptor):
...
+            validation.validate_single_setting(name, tk, manifest, s, value)
+        except TankError, e:
+            log.info("  ERROR - Parameter %s - Invalid value: %s" % (s,e))
+        else:
+            # validation is ok
+            if default is None:
+                # no default value
+                # don't report this
+                pass
+                #log.info("  Parameter %s - OK [no default value specified in manifest]" % s)
```
Comment (manneohrstrom) [inline python/tank/deploy/tank_commands/validate_config.py]: should this commented out line be removed?

File: python/tank/deploy/tank_commands/validate_config.py [hunk 17]
```diff
@@ -145,54 +145,54 @@ def _validate_bundle(log, tk, name, settings, descriptor):
...
+
+            elif manifest[s].get("type") == "hook" and value == "default":
+                # don't display when default values are used.
+                pass
+                #log.info("  Parameter %s - OK [using hook 'default']" % s)
```
Comment (manneohrstrom) [inline python/tank/deploy/tank_commands/validate_config.py]: should this commented out line be removed?

File: python/tank/deploy/tank_commands/validate_config.py [hunk 18]
```diff
@@ -145,54 +145,54 @@ def _validate_bundle(log, tk, name, settings, descriptor):
...
+
+            elif default == value:
+                pass
+                # don't display when default values are used.
+                #log.info("  Parameter %s - OK [using default value]" % s)
```
Comment (manneohrstrom) [inline python/tank/deploy/tank_commands/validate_config.py]: should this commented out line be removed?

File: python/tank/deploy/tank_commands/validate_config.py [hunk 19]
```diff
@@ -201,12 +201,13 @@ def _process_environment(log, tk, env):
         s = env.get_engine_settings(e)
         descriptor = env.get_engine_descriptor(e)
         name = "Engine %s / %s" % (env.name, e)
-        _validate_bundle(log, tk, name, s, descriptor)
+        _validate_bundle(log, tk, name, s, descriptor, engine_name=e)
```
Comment (manneohrstrom) [inline python/tank/deploy/tank_commands/validate_config.py]: I understand now how `engine_name` is used :) Nice!

File: python/tank/platform/bundle.py [hunk 20]
```diff
@@ -529,13 +529,28 @@ def __resolve_hook_path(self, settings_name, hook_expression):
             raise TankError("%s config setting %s: Configuration value cannot be None!" % (self, settings_name))
         
         path = None
+
+        # make sure to replace the engine token if it exists.
```
Comment (manneohrstrom) [inline python/tank/platform/bundle.py]: nitpicky: could this say 

```
# make sure to replace the {engine_name} token if it exists.
```

File: python/tank/platform/bundle.py [hunk 21]
```diff
@@ -804,28 +790,143 @@ def __post_process_settings_r(self, key, value, schema):
...
-        if value is None:
-            return value
-        
-        # try to get the type for the setting
-        # (may fail if the key does not exist in the schema,
-        # which is an old use case we need to support now...)
+
         try:
-            schema = self.__descriptor.get_configuration_schema().get(key)
+            engine_name = self.engine.name
```
Comment (manneohrstrom) [inline python/tank/platform/bundle.py]: feels like we should have note here (like in the code that was remoted) to highlight the fact that this is the `bundle` class and not all bundle instances have a `self.engine` - this is the reason we have to put that try/catch around it

File: python/tank/platform/bundle.py [hunk 22]
```diff
@@ -529,13 +529,28 @@ def __resolve_hook_path(self, settings_name, hook_expression):
...
+                raise TankError(
+                    "No engine could be determined for hook expression '%s'. "
+                    "The hook could not be resolved." % (hook_expression,))
+            else:
+                hook_expression = hook_expression.replace(
+                    constants.TANK_HOOK_ENGINE_REFERENCE_TOKEN,
+                    engine_name,
+                )
         
         # first the default case
```
Comment (manneohrstrom) [inline python/tank/platform/bundle.py]: This is the legacy case now, "old-style" hooks. I think that might be worth adding here for clarity

File: python/tank/platform/bundle.py [hunk 23]
```diff
@@ -544,28 +559,9 @@ def __resolve_hook_path(self, settings_name, hook_expression):
             #    parameters: [source_path, target_path]
```
Comment (manneohrstrom) [inline python/tank/platform/bundle.py]: can we get rid of the parameters line? This is old and not used anymore.

File: python/tank/platform/validation.py [hunk 24]
```diff
@@ -379,16 +391,38 @@ def validate(self, settings):
...
+            if settings_key in settings:
+                # value exists in the settings. use it.
+                settings_value = settings[settings_key]
+            else:
+                # Use the fallback default with an unlikely to be used value to
+                # detect cases where there is no default value in the schema.
+                no_default_value = constants.TANK_SCHEMA_NO_DEFAULT_VALUE_TEST_VALUE
+                settings_value = resolve_default_value(value_schema,
+                    no_default_value)
+                if settings_value == no_default_value:
```
Comment (manneohrstrom) [inline python/tank/platform/validation.py]: could `resolve_default_value` raise an exception instead of returning a magical token?

**Comment by manneohrstrom** [comment]
CR done! I have tried my best to think of what this change mean, but it's tricky! Overall, I think it looks totally awesome!

File: python/tank/deploy/tank_commands/validate_config.py (hunk 14 above)
Comment (josh-t) [inline python/tank/deploy/tank_commands/validate_config.py]: Done. Also added docstring for `_process_environment`.

File: python/tank/deploy/tank_commands/validate_config.py (hunk 15 above)
Comment (josh-t) [inline python/tank/deploy/tank_commands/validate_config.py]: now uses `manifest[s]` instead.

Yep, looks like you saw where the engine name was coming from in a later comment. It's just the paren't bundle in the config. I've since removed the `app_name` argument since it wasn't used. I'd included thinking it would be needed, but it isn't.

File: python/tank/deploy/tank_commands/validate_config.py (hunk 16 above)
Comment (josh-t) [inline python/tank/deploy/tank_commands/validate_config.py]: These commented out lines are in the existing code. I just outdented this whole block so it shows up as modified. I'll go through and remove these.

File: python/tank/platform/bundle.py (hunk 20 above)
Comment (josh-t) [inline python/tank/platform/bundle.py]: Done!

File: python/tank/platform/bundle.py (hunk 22 above)
Comment (josh-t) [inline python/tank/platform/bundle.py]: I modified the comment.

File: python/tank/platform/bundle.py (hunk 21 above)
Comment (josh-t) [inline python/tank/platform/bundle.py]: Copied the removed comment here for clarity. :)

File: python/tank/platform/bundle.py (hunk 23 above)
Comment (josh-t) [inline python/tank/platform/bundle.py]: done.

File: python/tank/deploy/tank_commands/console_utils.py (hunk 12 above)
Comment (josh-t) [inline python/tank/deploy/tank_commands/console_utils.py]: Makes sense!

File: python/tank/platform/validation.py [hunk 25]
```diff
@@ -379,16 +391,39 @@ def validate(self, settings):
...
+            if settings_key in settings:
+                # value exists in the settings. use it.
+                settings_value = settings[settings_key]
+            else:
+                # Use the fallback default with an unlikely to be used value to
+                # detect cases where there is no default value in the schema.
+                try:
+                    settings_value = resolve_default_value(value_schema,
+                        raise_if_missing=True)
+                except TankError:
```
Comment (manneohrstrom) [inline python/tank/platform/validation.py]: wondering if this should be a special exception rather than TankError? Right now, we assume that this TankError means that the method could not identify a default value, but I am guessing there could be other reasons a tank error may be raise?

Comment (josh-t) [inline python/tank/platform/validation.py]: I don't think it'll raise anything else currently, but I think you're right. I've added a new exception to `tank.errors` and used it instead.
